### ISSUE TITLE
Add support for RemoteCertificateValidationCallback with ClientWebSocket

### ISSUE THREAD

**Issue by BravoTango86** [dotnet/corefx#12038]
There is no way of overriding the validation of server certificates on an SslStream utilised by ClientWebSocket.

Could this be exposed through ClientWebSocketOptions?

_EDIT 3/5/2018 by @stephentoub. Added proposal:_
```C#
public sealed class ClientWebSocketOptions
{
    public RemoteCertificateValidationCallback RemoteCertificateValidationCallback { get; set; }
    ...
}
```

**Comment by davidsh** [comment]
cc: @CIPop @stephentoub

**Comment by CIPop** [comment]
> Could this be exposed through ClientWebSocketOptions?

I think that's the right place to put the new API too. The API should be identical to one on HttpClient:
```C#
        public System.Func<System.Net.Http.HttpRequestMessage, System.Security.Cryptography.X509Certificates.X509Certificate2, System.Security.Cryptography.X509Certificates.X509Chain, System.Net.Security.SslPolicyErrors, bool> ServerCertificateCustomValidationCallback { get { throw null; } set { } }
```

For the API proposal please follow https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/adding-api-guidelines.md.
If anyone picks this item, please focus on the ManagedWebSocket implementation initially.

**Comment by qmfrederik** [comment]
@stephentoub I've again hit a use case where I'm wanting to use a ClientWebSocket with client certificates, and where the client certificates come from a private CA (a Kubernetes cluster in this case).

I don't necessarily want to add that CA to the list of trusted CAs, but rather manage the validation in code.

In https://github.com/dotnet/corefx/issues/5120#issuecomment-353613941 you said to open an API issue to support certificate validation in ClientWebSocket; it looks like this issue is just about that.

So, could this issue serve as the API proposal issue? The API proposal from @CIPop seems to make sense:

```diff
ClientWebSocket
{
 +       public Func<HttpRequestMessage, X509Certificate2, X509Chain, SslPolicyErrors, bool> ServerCertificateCustomValidationCallback { get; set; }
}
```

**Comment by stephentoub** [comment]
We can use this issue for it, but the proposed API would actually need to be on ClientWebSocketOptions rather than on ClientWebSocket.  I also think it would behoove us to instead consider something like https://github.com/dotnet/corefx/issues/26567 rather than adding lots of individual properties as we find more and more areas of SslStream configuration that's desirable to be exposed.

**Comment by tintoy** [comment]
@stephentoub - if I'm reading the [documentation](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/adding-api-guidelines.md) correctly, the way to proceed with a proposal is to fork the corefx repo and actually make the proposed changes so they can be reviewed and discussed? Have I got that right?

If so, I'd be willing to put my hand up to have a go at helping to move this forward :-)

**Comment by stephentoub** [comment]
> the way to proceed with a proposal is to fork the corefx repo and actually make the proposed changes so they can be reviewed and discussed? Have I got that right?

That's not necessary; we just need a solid proposal for what the API would look like.

**Comment by tintoy** [comment]
Ok I'm still happy to pitch in FWIW (I can look at older proposals for guidance). Just want to check that you're ok with me making the attempt (if someone else is already working on it I wouldn't want to double up on effort is all).

**Comment by stephentoub** [comment]
Actually, I take back what I said about using SslClientAuthenticationOptions:
1. The ClientWebSocketOptions already exposes ClientCertificates, and it'd be a bit strange to have both that and an SslOptions that also had ClientCertificates (though we could smooth over that a bit by just having the former delegate to the latter).
2. ClientWebSocket isn't necessarily tied in implementation to System.Net.Security and SslStream, so it could be strange in the public API to use a type that is.

So, I suggest we just keep this simple and add:
```C#
public sealed class ClientWebSocketOptions
{
    public RemoteCertificateValidationCallback RemoteCertificateValidationCallback { get; set; }
    ...
}
```
It'd be nice to get this into 2.1 if possible.
cc: @geoffkizer, @davidsh

**Comment by davidsh** [comment]
I agree with this API approach with just adding this API to the existing `ClientWebSocketOptions` class:

```c#
public RemoteCertificateValidationCallback RemoteCertificateValidationCallback { get; set; }
```

**Comment by brendandburns** [comment]
Would love to see this get added!

**Comment by stephentoub** [comment]
@davidsh, I just looked at the WinRT implementation, and it appears to be based on MessageWebSocket, which has an API surface that I'm not sure could support this.  Any concerns?

**Comment by davidsh** [comment]
In that case, it would have to throw PNSE on UAP for now.  Long-term we should consider moving the UAP version to use the "managed" implementation, same as the rest of CoreFx.

**Comment by karelz** [comment]
@stephentoub can you please ping shiproom with brief description of the API addition? Given we are quite past API freeze, we should do it.

**Comment by stephentoub** [comment]
> Given we are quite past API freeze, we should do it.

What is the purpose of the API freeze? a) This was approved before it. b) This is making something public that was already implemented before it.

**Comment by karelz** [comment]
Good questions, I believe the main desire to track what's being added at this point - from partner impact point of view (e.g. Xamarin) and risk assessment & focus. Double check with them.
The API was approved after API freeze, which was on 3/1 😉.

**Comment by qmfrederik** [comment]
From a partner impact point of view, the .NET client for Kubernetes (see @brendandburns 's message earlier) would benefit hugely from this still making it to 2.1; the alternative is duplicating the managed WebSocket client in the Kubernetes client which is not a very attractive option.

**Comment by karelz** [comment]
I've send email to shiproom, let's see if we get it approved.

**Comment by anurse** [comment]
FYI, this affects SignalR customers as well. We'd definitely like this fix as soon as possible.

**Comment by karelz** [comment]
We've got shiproom approval, to get it in 2.1 - @stephentoub do you want to reopen your PR and push it through?

**Comment by karelz** [comment]
FYI @marek-safar API changes after API freeze, not sure if it impacts you ...

**Comment by karelz** [comment]
Thanks @stephentoub for pushing it through and thanks everyone for impact details, it's appreciated.

**Comment by gcadmes** [comment]
So after reading through this thread, It's safe to assume that the new API will be available in netstandard2.1, yes?
May I ask when that might go primetime?

**Comment by mitchcapper** [comment]
It is merged into master at this point so yes I am quite confident it should be in 2.1.   There should be a new beta of 2.1 any day now, depending when they flighted it it might  be in that beta if so I will add support quickly.   Keep in mind you would need to roll your own beta of 2.1 to use it however.    Official 2.1 release I think it slated for H1.

**Comment by ygoe** [comment]
@karelz So if there's no netstandard21, unfortunately this new feature isn't accessible to my library. Should I use reflection to see if the runtime has that property, and set it?

**Comment by ygoe** [comment]
@danmosemsft Will there be a netstandard21 or something that includes this feature sometime? Or is this cross-targeting meant for a longer time?

**Comment by danmosemsft** [comment]
@ygoe there will certainly be more versions of .NET Standard. We [announced ](https://blogs.msdn.microsoft.com/dotnet/2018/05/07/net-core-3-and-support-for-windows-desktop-applications/) there will be an update next year aligned with .NET Core 3.0. 

When we update it depends on how quickly other platforms (mainly Mono/Xamarin) pick up new API's. There's more discussion [here](https://github.com/dotnet/standard/issues/682) 

In short, multitargeting should be temporary as this API should end up in the next .NET Standard.

**Comment by ygoe** [comment]
Just a side note, [Visual Studio team has just decided](https://developercommunity.visualstudio.com/content/problem/82419/platform-context-cannot-be-selected-in-code-editor.html) that multi-targeting is not a priority and that my bad experience with it is acceptable. So I have decided that multi-targeting is not an option. My reflection-based implementation seems to do the job without all the hassle.
